lib/kube/proxy/testing/kube_server/kube_mock.go

```diff
@@ -527,3 +538,82 @@ func (s *KubeMockServer) selfSubjectAccessReviews(w http.ResponseWriter, req *ht
 
 	return s1, nil
 }
+
+// portforward supports SPDY protocols only. Teleport always uses SPDY when
+// portforwarding to upstreams even if the original request is WebSocket.
+func (s *KubeMockServer) portforward(w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	_, err = httpstream.Handshake(req, w, []string{PortForwardProtocolV1Name})
+	if err != nil {
+		return
+	}
+
+	streamChan := make(chan httpstream.Stream, 1)
+
+	upgrader := spdystream.NewResponseUpgraderWithPings(defaults.HighResPollingPeriod)
+	conn := upgrader.UpgradeResponse(w, req, httpStreamReceived(req.Context(), streamChan))
+	if conn == nil {
+		err = trace.ConnectionProblem(nil, "unable to upgrade SPDY connection")
+		return
+	}
+	defer conn.Close()
+	var (
+		data      httpstream.Stream
+		errStream httpstream.Stream
+	)
+
+	for {
+		select {
+		case <-conn.CloseChan():
+			return
+		case stream := <-streamChan:
+			streamType := stream.Headers().Get(StreamType)
+			switch streamType {
+			case StreamTypeError:
+				errStream = stream
+			case StreamTypeData:
+				data = stream
+			}
+		}
+		if errStream != nil && data != nil {
+			break
+		}
+	}
+
+	buf := make([]byte, 1024)
+	n, err := data.Read(buf)
+	if err != nil {
+		errStream.Write([]byte(err.Error()))
+	}
+	data.Write([]byte(fmt.Sprintf(PortForwardPayload, p.ByName("podName"), string(buf[:n]))))
```

Comment:
I'd rather write `PortForwardPayload` explicitly here, makes it easier to reason about it when reading.